framework/src/meshgenerators/PatchMeshGenerator.C

```diff
@@ -0,0 +1,381 @@
+                             "valid in 2 or 3 dimensions.");
+
+  MooseEnum elem_types("QUAD4 QUAD8 HEX8 HEX20", "QUAD4");
+  params.addParam<MooseEnum>("elem_type",
+                             elem_types,
+                             "The type of element from libMesh to "
+                             "generate (default: linear element for "
+                             "requested dimension)");
+  params.addParam<Real>("x_length", 0.24, "Length of the domain in the x direction.");
+  params.addParam<Real>("y_length", 0.12, "Length of the domain in the y direction.");
+  params.addParam<Real>("z_length", 0.0, "Length of the domain in the z direction.");
+  params.addParam<Real>("x_offset", 0.0, "Offset of the Cartesian origin in the x direction.");
+  params.addParam<Real>("y_offset", 0.0, "Offset of the Cartesian origin in the y direction.");
+  params.addParam<Real>("z_offset", 0.0, "Offset of the Cartesian origin in the z direction.");
+  return params;
+}
+
+PatchMeshGenerator::PatchMeshGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _dim(getParam<MooseEnum>("dim")),
+    _elem_type(getParam<MooseEnum>("elem_type")),
+    _xlength(getParam<Real>("x_length")),
+    _ylength(getParam<Real>("y_length")),
+    _zlength(getParam<Real>("z_length")),
+    _xoffset(getParam<Real>("x_offset")),
+    _yoffset(getParam<Real>("y_offset")),
+    _zoffset(getParam<Real>("z_offset"))
+{
+  if (_xlength <= 0.0)
+    paramError("x_length", "Must be greater than zero");
+
+  if (_ylength <= 0.0)
+    paramError("y_length", "Must be greater than zero");
+
+  if (_dim == 2 && (_elem_type == "HEX8" || _elem_type == "HEX20"))
+    paramError("elem_type", "Must be QUAD4 or QUAD8 for 2-dimensional meshes.");
+
+  if (_dim == 2 && _zlength != 0.0)
+    paramError("z_length", "Must be zero for 2-dimensional meshes");
+
+  if (_dim == 3 && !parameters.isParamSetByUser("elem_type"))
+    _elem_type = "HEX8";
+
+  if (_dim == 3 && (_elem_type == "QUAD4" || _elem_type == "QUAD8"))
+    paramError("elem_type", "Must be HEX8 or HEX20 for 3-dimensional meshes.");
+
+  // If domain dimensions are not set by user for 3-dimensions a unit cube is used
+  // as default as per MacNeal and Harder (1985)
+  if (_dim == 3)
+  {
+    if (!parameters.isParamSetByUser("x_length"))
+      _xlength = 1.0;
+    if (!parameters.isParamSetByUser("y_length"))
+      _ylength = 1.0;
+    if (!parameters.isParamSetByUser("z_length"))
+      _zlength = 1.0;
+  }
+
+  if (_dim == 3 && _zlength <= 0.0)
+    paramError("z_length", "Must be greater than zero for 3-dimensional meshes");
+}
+
+std::unique_ptr<MeshBase>
+PatchMeshGenerator::generate()
+{
+  std::unique_ptr<ReplicatedMesh> mesh = libmesh_make_unique<ReplicatedMesh>(comm(), 2);
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+
+  unsigned num_nodes = 0;
+  if (_elem_type == "QUAD4")
+    num_nodes = 8;
```

Comment:
`num_nodes` is a static constant available for each element type in libMesh. You shouldn't be supplying these numbers directly here.

Reply:
This doesn't refer to the num_nodes in a single QUAD4.  This refers to the total number of number of nodes present in the patch mesh if it is made up of QUAD4s.  Similarly for the other element types.

Reply:
Ah ok, Then this is fine.